### DISCUSSION TITLE
Can PO number be created from pattern?

### DISCUSSION THREAD

**Discussion by miggland** [inventree/InvenTree#3163] [Ideas]
For reasons, my company creates purchase orders with a pattern of four digits followed by a dash, and then a second part consisting of four digits.
The first four digits are the part that should be incremented for each order, while the second part refers to which project they belong to - ie its manually added.
I've been creating POs in InvenTree, and it is smart - unfortunately, it increments the project number and not the first part, so I have to manually check what the correct number is each time.

Is there some way in which we could create the PO number from a pattern, ie define:
po_pattern = "nnnn-blah"
and then have InvenTree realise it should increment the nnnn part, and just insert the rest?

Maybe this would open for creating PO numbers from other variables, such as dates, as well?

Edit: @SchrodingersGat - I have opened an issue here - https://github.com/inventree/InvenTree/issues/3167

**Comment by matmair** [comment]
Would be a great addition. We already have something some name templating implemented - that work could be generalized and reused probably.

See:
- https://inventree.readthedocs.io/en/latest/report/labels/#context-data
- https://inventree.readthedocs.io/en/latest/settings/global/#parts part display seems like it could be usefull @miggland 


Similar thoughts were brought up several times - although I am not aware of accessible forks or plugins for this. See:
- https://github.com/inventree/InvenTree/issues/2848
- #708

**Reply by matmair** [reply]
If this gains some interest I would appreciate a FR so we can track all requirements and parameters following the template

**Comment by SchrodingersGat** [comment]
@miggland I would definitely support this. We have facility for "templating" various fields, although it is not immediately obvious to me how to implement your exact use case. 

1. We would need some way to signify which part of the reference is "fixed" and which increments
2. We need a way to find the "highest" number (based on the incrementing portion of the field) and generate the "next" one according to some rule
3. We need to be able to "sort" reference fields naturally

Regarding 3. we keep a "shadow copy" of the reference field for each order (PO / SO / BO) which is an integer cast of the reference:

https://github.com/inventree/InvenTree/blob/79f498a648e889454e083023ae003c96e359cb8c/InvenTree/InvenTree/models.py#L96

We attempt to create an integer version of each reference which is stored in the database. This is necessary for efficient natural sorting of  orders - otherwise string sort has to be used.

We would be happy to support a PR for this, but it requires some careful consideration regarding implementation

**Reply by matmair** [reply]
@SchrodingersGat should we make a FR out of this to discuss the architecture? I have some ideas. Maybe this could be implemented model-wide as a generic mixin and rolled out across all orders, company and customer. Controlled via a custom Meta class.
We could then just create a reference implementation - just test that and keep testing time low that way.

**Comment by miggland** [comment]
Thanks for taking this up! I'll continue here, because the FR should be more for detailed discussions on implementation I guess.

At the moment, I manually put a number like '1043-6790' in the PO reference. The next one (assuming the same project) would be '1044-6790'. However, InvenTree puts '1043-6791'. It is smart enough to increment the number already, it just increments the wrong part.

As you write in the FR @SchrodingersGat , regex can be confusing! I was thinking more along the lines of (which I have seen in other software, eg my scanner:

```
Filename
(Max.64 characters)
 Example:
 Image#n. The sequential number is given to the place of "#n".
 Image#d. Date is given to the place of "#d". 
```
The user could fill in a reference template like:
`PO #NNNN`
to get "PO1234" and "PO1235" ...
or
`This is your #N order`
as a stupid example to get "This is your 3 order"
or 
`PO #YYYY-#NNN` to get
"PO2022-456" and "PO2022-457".
For may case, I would use:
`PO #NNNN-xxxx`

to have "PO1234-xxxx" show up in the form, and I would fix the project number manually.

We would never know how exactly users want the number part to behave. Eg in the third example, what does a user want next year? Start at 1, continue, ...?

Originally, I wasn't talking about enforcement either - just pre-filling the form. I guess in the API one would still have to supply value though?
